### ISSUE TITLE
Alpha 5, V20, browsing for CRT image and changing folders results in a Fatal Error

### ISSUE THREAD

**Issue by ThreepwoodLeBrush** [MJoergen/C64MEGA65#37]
**Description:**
For a collection of CRT-files I use a folder structure on my SD of C64/CRTONL64/[First Letter]. When I enter such a Letter Subfolder, even one with little files in there, and then return to the parent directory ".." I get this crash in V20 (I had no such issue in V19):

**Error Message:**
FATAL ERROR:

Corrupt memory structure:
Linked-list boundary

Core stopped. Please reset the machine.

![image](https://user-images.githubusercontent.com/59653481/236866267-e395d906-f562-4faf-998c-21ba2b727e31.png)

**Steps to replicate:**
1. Open menu via HELP and choose "CRT: Load"
2. Traverse directory hierarchy down into C64/CRTONL64/[Number|Letter]
3. Go up one level by choosing "..", this will trigger the crash.

An image of my MEGA65 SD-Card has been provided to @sy2002. Note that the core only crashes when going all the way down into a single letter folder. One level higher everything works and no crash occurs.

**Comment by sy2002** [comment]
Thank you for reporting this one. Your support is highly appreciated.

**Comment by ThreepwoodLeBrush** [comment]
With V5 A21 I cannot reproduce this error anymore.

I use Kibo's network enabled core for sending files over and sent the new core file, the c64mega65 config file as well as the CRT of Eye of the Beholder over.

After flashing the new core A21 and trying to reproduce it, the error no longer appears.

**Comment by sy2002** [comment]
@ThreepwoodLeBrush Super scary! Between Alpha 20 and Alpha 21 we did not change a single line of code when it comes to the FAT32 library (software) and/or the SD card controller (hardware).

We did change something between Alpha 19 and Alpha 20 though, i.e. we fixed this bug here:
https://discord.com/channels/719326990221574164/794775503818588200/1104211122371952730

We fixed the bug by doing this here:
https://github.com/sy2002/QNICE-FPGA/commit/c531c8d691a96210b93b759b5dc4dae41295cbbc

So my "hope" was, that this is a stable bug situation (i.e. well reproducible bug) that kind of is/was a regression from the bugfix that happened between A19 and A20.

If you now say, that A21 "fixes" it (or you at least cannot reproduce it), then the situation is much more complex as this seems to be a sporadic bug: On a high level, three possible reasons:

1. Corrupted FAT32 file system, i.e. running fsck on Linux/Mac or running chkdsk under windows would fix it.

2. Bug in our FAT32 library that is only triggered in a certain edge case

3. Bug in our hardware implementation (VHDL side of things)

4. Corrupted SD card

Hard to debug. I will try to reproduce using the binary dump of your SD card that you've sent me via Discord DM.

**Comment by sy2002** [comment]
This is only an FYI because I will also try to reproduce directly on the MEGA65 hardware.

I downloaded the binary SD card clone successfully and was able to successfully mount it. But the mounting daemon claimed "corruption", "danger and potential harm for your system" when mounting it. I continued and mounted anyway.

To check/analyze if above-mentioned "theory #1" might be valid, I ran the FAT32 version of fschk. Here is the output log:

```bash
TITANIUM-M5:monitor $ fsck_msdos -y /dev/rdisk5s1
** /dev/rdisk5s1 (NO WRITE)
** Phase 1 - Preparing FAT
FAT[0] is incorrect (is 0x0; should be 0xFFFFFF8)
Correct? no
** Phase 2 - Checking Directories
`..' entry in /D64 has non-zero start cluster
Correct? no
`..' entry in /D81 has non-zero start cluster
Correct? no
** Phase 3 - Checking for Orphan Clusters
Found orphan cluster(s)
Fix? no
Found 104 orphaned clusters
Free space in FSInfo block (7201926) not correct (7203421)
Fix? no
2194 files, 28813684 KiB free (7203421 clusters)
TITANIUM-M5:monitor $ 
```

So there is something "non-optimal" going on with this SD card clone image. Next step is to use an unmodified version of the binary image that I downloaded and put it on a brand new SD card (binary clone) and test it in the MEGA65 running A20 and A21.

**Comment by ThreepwoodLeBrush** [comment]
> Maybe two additional question @ThreepwoodLeBrush to rule out workflow-related issues:
> 
>     1. Did you change anything on your SD card between 20 and 21, for example clean it up (delete files/folders/anything) before you made the binary clone for me or things like this?
> 
>     2. What exactly is "Kibo's network enabled core for sending files over"?

1. Between 20 and 20 I copied over the new 21 core file, the c64mega65 config file and the CRT of "Eye of the Beholder". Nothing else

2. On the MEGA65 Discord the user "Kibo" is working on a MEGA65 core that works with megaftp via ethernet, I FTPed the files over. This is still beta and the transfer often breaks, requiring killing megaftp and retransferring the files, which is probably the culprit.

After this copy operation, I cannot reproduce the error in neither A20, nor A21, so maybe that file transfer fixed the FAT? It is possible that using megaftp via ethernet corrupted the filesystem.

Since I can no longer reproduce the error, to me this issue can be chalked off as an SD card issue and closed. Thank you for looking into it!

**Comment by sy2002** [comment]
@ThreepwoodLeBrush thank you for the update. For making sure that we are on the safe side, here is what I did:

1. Cloned the bitwise image that you've sent me on a brand new 32 GB Sandisk SD card.
2. Tested the above-mentioned Steps to replicate with *Alpha 20*, I literally went through all of the "letters" and I tested the SD card in the bottom tray as well as in the tray on the back side
3. Repeat the previous step with *Alpha 21*

The result: I cannot reproduce it, too.

So I am closing this issue.
